C++17 rewrite

PeterMatula opened the pull request:
C++17 deprecated and removed some language features. This made this project not compile inside [RetDec](https://github.com/avast/retdec).

I replaced these features and bumped Travis and Appveyor environments. I also bumped stuff in `README.md` and `CHANGELOG.md`.

Now it compiles without warnings or errors.

Since PeLib does not have its own unit or regression tests, can one (both?) of you (@s3rvac, @metthal) please look at it? I was careful, but it is easy to mess up those lambdas :-)

I will need this for https://github.com/avast/retdec/issues/650. But for now, I will use reference to this branch (instead of `master`) in RetDec, so it is not a total blocker. It will also run RetDec tests with these changes, so it may catch some problems before we merge here.

PeterMatula committed:
CMakeLists.txt: set C++ standard to 17 and require it

PeterMatula committed:
appveyor+travis: temp enable builds of c++17 branch

PeterMatula committed:
appveyor+travis: bump build environments fot c++17

PeterMatula committed:
replace uses of std::bind2nd and std::mem_fun_ref

These were deprecated or removed in C++17.

PeterMatula committed:
.travis.yml: do not install gcc-multilib on linux

PeterMatula committed:
README+CHANGELOG: bump versions, mention C++17, etc.

PeterMatula committed:
appveyor+travis: remove temp triggers for c++17 branch

PeterMatula committed:
ImportDirectory.h: fix std::bind() use (maybe)

s3rvac commented on README.md:
```diff
@@ -22,8 +22,8 @@ target_link_libraries(project-that-needs-pelib pelib)
-* A compiler supporting C++14
-  * On Windows, only Microsoft Visual C++ is supported (version >= Visual Studio 2015).
+* A compiler supporting C++17
+  * On Windows, only Microsoft Visual C++ is supported (version >= Visual Studio 2017 version 15.7).
```
Shouldn't we require version 15.9, which is the latest version of MSVC 2017? MSVC 2017 should be C++17-feature-complete since 15.8, so I suggest requiring at least 15.8.

PeterMatula replied:
I don't know about 15.9 - I'm not using VS and I don't know if it is common to upgrade, or if there are a lot of people with older versions - I did not want to be too strict for this reason.
I was searching a long time for the C++17 version to require. 15.7 came up the most. Then, it was mentioned at wikipedia page for C++17 (https://en.wikipedia.org/wiki/C%2B%2B17#Compiler_support), so I just used that :-)
I have no problem to bump this, but I fear that a lot of people start writing to us about using older versions.

s3rvac replied:
I have based my comment on [compiler support on cppreference](https://en.cppreference.com/w/cpp/compiler_support). Version 15.8 (19.15 in there, see [this page](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warnings-by-compiler-version?view=vs-2019) for a mapping) should have nearly full support of C++ 17, from both language and library viewpoints. However, even 15.7 is fine for our purposes, so you can keep this version there if you want to. Alternatively, we do not have to specify it all all (just say that we require Visual Studio 2017 or later).

s3rvac commented on include/pelib/ImportDirectory.h:
```diff
@@ -63,11 +63,19 @@ namespace PeLib
+						auto Iter = std::find_if(
+								FileIter->originalfirstthunk.begin(),
+								FileIter->originalfirstthunk.end(),
+								std::bind(comp, std::placeholders::_1, value)
```
I suggest using lamdas instead of `std::bind()` in C++17 code. For more information, please see _"Item 34: Prefer lambdas to std::bind"_ in [Scott Meyers: Effective Modern C++](https://www.amazon.com/gp/product/1491903996/).

This note concerns several pieces of code in the present PR.

PeterMatula replied:
In those other cases [where `std::bind()` is used], I was simply not able to replace it with something else. I bet it is possible, but those `ResComparer<T>` are weird and I got lost in some crazy compiler errors.
In this case you commented on, there was actually a lambda originally. I changed it last minute before the commit because I thought `std::bind()` expresses better what is going on. In other cases, there is usually a call to method of the object you get from lambda. Here, this object is used as one of the parameters of `comp` function - so I thought this is expressing that better.
Are reasons for preferring lambdas mentioned in that book applicable to this case?

PeterMatula replied:
The problem with those comparators is, that it is doing overloading in that single function that uses it - the correct comparator (string/int) is selected based on the type of ID in parameter.

s3rvac replied:
In general, lambdas are more readable, more expressive, and may be more efficient than `std::bind` (see the referenced book). In this particular case, I would get rid of that comparators and use overloaded member functions. In my opinion, it makes the code simpler and easier to understand. [Here is a patch](https://files.petrzemek.net/tmp/Replace-comparators-with-overloaded-member-functions.patch) that you can apply with `git am < file.patch` (I was unable to upload it here as GitHub failed with "something went really wrong").

PeterMatula replied:
Thanks, I applied the patch. It is much better now. I was afraid to touch those comparator structures, but it looks ok in your code.

s3rvac reviewed (approved):
LGTM. I have included two remarks to consider.

PeterMatula committed:
PeHeader.h: fix usage of comp function in std::max_element() lambdas

PeterMatula commented:
LOL 651987a

I thought there is something fishy about this when I wrote it. But it "compiled" so I thought ok, probably some _partial application_ is going on in there - that method has one argument but was called with zero.
It turns out, it did not compile when used in RetDec - templates :-)
I guess no _partial application_ in C++.
Is it ok now? That comparator object description for `std::max_element()` is non-trivial.

s3rvac commented:
> Is it ok now? That comparator object description for std::max_element() is non-trivial.

The comparison function should take two arguments and return `true` when the first one is smaller than the second one. The current code is
```cpp
[](const auto& i1, const auto& i2) { return i1.biggerFileOffset(i2); }
```
Member function `biggerFileOffset()` is
```cpp
bool PELIB_IMAGE_SECTION_HEADER::biggerFileOffset(const PELIB_IMAGE_SECTION_HEADER& ish) const
{
    return PointerToRawData < ish.PointerToRawData;
}
```
So, the comparator is correct.

As for the compilation, when a compiler encounters a template, it first only checks its syntax. It checks its semantic only after the template is instantiated with particular types/values.

PeterMatula commented:
Thanks. I will merge this once RetDec tests that use this pass.

s3rvac committed:
Replace comparators with overloaded member functions.

This simplifies the code and allows us to use more readable lambdas than
std::bind().

PeterMatula commented:
Everything looks good now, but I will still wait for RetDec regression tests to pass witch this branch - I'm currently running this code inside RetDec's `issue-650` (C++17 support) branch and there are some more complications in there - likely unrelated to changes here.
